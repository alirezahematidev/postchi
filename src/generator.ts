import path from 'node:path';
import fs from 'fs-extra';
import pc from 'picocolors';
import type { PostchiConfig } from './types/config';
import type { PostmanCollection, PostmanItem, PostmanRequest } from './types/postman';

export class ApiGenerator {
  private collection: PostmanCollection;
  private config: Required<PostchiConfig>;

  constructor(collectionPath: string, config: Required<PostchiConfig>) {
    this.collection = JSON.parse(fs.readFileSync(collectionPath, 'utf-8'));
    this.config = config;
  }

  private generateTypeName(name: string): string {
    return name
      .split(/[^a-zA-Z0-9]/)
      .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
      .join('');
  }

  private generateRequestType(request: PostmanRequest): string {
    const typeName = this.generateTypeName(request.url.raw);
    let typeDefinition = `export interface ${typeName}Request {\n`;

    // Add query parameters
    if (request.url.query) {
      for (const query of request.url.query) {
        typeDefinition += `  ${query.key}: string;\n`;
      }
    }

    // Add body type if present
    if (request.body?.raw) {
      try {
        const bodyType = JSON.parse(request.body.raw);
        typeDefinition += `  body: ${JSON.stringify(bodyType, null, 2)
          .split('\n')
          .map((line) => '  ' + line)
          .join('\n')};\n`;
      } catch {
        typeDefinition += `  body: any;\n`;
      }
    }

    typeDefinition += '}\n\n';
    return typeDefinition;
  }

  private generateJavaScriptRequestType(request: PostmanRequest): string {
    // For JavaScript, we'll generate JSDoc comments instead of TypeScript interfaces
    const typeName = this.generateTypeName(request.url.raw);
    let typeDefinition = `/**\n * @typedef {Object} ${typeName}Request\n`;

    // Add query parameters
    if (request.url.query) {
      for (const query of request.url.query) {
        typeDefinition += ` * @property {string} ${query.key}\n`;
      }
    }

    // Add body type if present
    if (request.body?.raw) {
      try {
        const bodyType = JSON.parse(request.body.raw);
        typeDefinition += ` * @property {Object} body\n`;

        for (const [key, value] of Object.entries(bodyType)) {
          const type = typeof value;
          typeDefinition += ` * @property {${type}} body.${key}\n`;
        }
      } catch {
        typeDefinition += ` * @property {*} body\n`;
      }
    }

    typeDefinition += ` */\n\n`;
    return typeDefinition;
  }

  private generateFetchFunction(request: PostmanRequest, name: string): string {
    const typeName = this.generateTypeName(request.url.raw);
    const functionName = name.toLowerCase().replace(/[^a-zA-Z0-9]/g, '_');

    let functionDefinition = '';
    if (this.config.language === 'typescript') {
      functionDefinition = `export async function ${functionName}(`;

      // Add parameters
      const params: string[] = [];
      if (request.url.query) {
        params.push(`params: ${typeName}Request`);
      }
      if (request.body?.raw) {
        params.push(`body: ${typeName}Request['body']`);
      }
      functionDefinition += params.join(', ');
      functionDefinition += `): Promise<Response> {\n`;
    } else {
      functionDefinition = `/**\n`;
      if (request.url.query) {
        functionDefinition += ` * @param {${typeName}Request} params\n`;
      }
      if (request.body?.raw) {
        functionDefinition += ` * @param {Object} body\n`;
      }
      functionDefinition += ` * @returns {Promise<Response>}\n */\n`;
      functionDefinition += `export async function ${functionName}(`;
      const params: string[] = [];
      if (request.url.query) {
        params.push(`params`);
      }
      if (request.body?.raw) {
        params.push(`body`);
      }
      functionDefinition += params.join(', ');
      functionDefinition += `) {\n`;
    }

    // Build URL
    functionDefinition += `  const url = new URL('${request.url.raw}');\n`;
    if (request.url.query) {
      functionDefinition += `  Object.entries(params).forEach(([key, value]) => {\n`;
      functionDefinition += `    url.searchParams.append(key, value);\n`;
      functionDefinition += `  });\n`;
    }

    // Add headers
    functionDefinition += `  const headers = new Headers({\n`;

    for (const header of request.header) {
      functionDefinition += `    '${header.key}': '${header.value}',\n`;
    }

    functionDefinition += `  });\n`;

    // Add request implementation based on the selected handler
    if (this.config.requestHandler === 'fetch') {
      // Fetch implementation
      functionDefinition += `  return fetch(url.toString(), {\n`;
      functionDefinition += `    method: '${request.method}',\n`;
      functionDefinition += `    headers,\n`;
      if (request.body?.raw) {
        functionDefinition += `    body: JSON.stringify(body),\n`;
      }
      functionDefinition += `  });\n`;
    } else if (this.config.requestHandler === 'axios') {
      // Axios implementation
      functionDefinition += `  return axios({\n`;
      functionDefinition += `    method: '${request.method.toLowerCase()}',\n`;
      functionDefinition += `    url: url.toString(),\n`;
      functionDefinition += `    headers: Object.fromEntries(headers.entries()),\n`;
      if (request.body?.raw) {
        functionDefinition += `    data: body,\n`;
      }
      functionDefinition += `  });\n`;
    }

    functionDefinition += `}\n\n`;
    return functionDefinition;
  }

  private processItem(item: PostmanItem, types: string[], functions: string[]): void {
    if (item.request) {
      if (this.config.language === 'typescript') {
        types.push(this.generateRequestType(item.request));
      } else {
        types.push(this.generateJavaScriptRequestType(item.request));
      }
      functions.push(this.generateFetchFunction(item.request, item.name));
    }

    if (item.item) {
      for (const subItem of item.item) {
        this.processItem(subItem, types, functions);
      }
    }
  }

  public generate(): void {
    const types: string[] = [];
    const functions: string[] = [];

    console.info(
      pc.blue('üîç Processing Postman collection:'),
      pc.cyan(this.collection.info?.name || 'Unnamed collection'),
    );

    for (const item of this.collection.item) {
      this.processItem(item, types, functions);
    }

    // Create output directory if it doesn't exist
    if (!fs.existsSync(this.config.output)) {
      console.info(pc.blue('üìÅ Creating output directory:'), pc.cyan(this.config.output));
      fs.mkdirSync(this.config.output, { recursive: true });
    }

    if (this.config.strategy === 'single-file') {
      // Generate a single file with all types and functions
      const output = [
        '// Generated API client',
        '// Do not modify this file manually',
        '',
        // Add axios import if needed
        this.config.requestHandler === 'axios' ? "import axios from 'axios';\n" : '',
        ...types,
        ...functions,
      ].join('\n');

      const extension = this.config.language === 'typescript' ? 'ts' : 'js';
      const outputFilePath = path.join(this.config.output, `api-client.${extension}`);

      console.info(pc.blue('üìù Writing API client to:'), pc.cyan(outputFilePath));
      fs.writeFileSync(outputFilePath, output);
    } else if (this.config.strategy === 'multi-file') {
      // Generate separate files for types and functions
      const typesOutput = [
        '// Generated API types',
        '// Do not modify this file manually',
        '',
        ...types,
      ].join('\n');

      const functionsOutput = [
        '// Generated API functions',
        '// Do not modify this file manually',
        '',
        // Add axios import if needed
        this.config.requestHandler === 'axios' ? "import axios from 'axios';\n" : '',
        // Import types if using TypeScript
        this.config.language === 'typescript'
          ? 'import { ' +
            types
              .filter((type) => type.startsWith('export interface'))
              .map((type) => type.split(' ')[2].split('Request')[0] + 'Request')
              .join(', ') +
            " } from './api-types';\n"
          : '',
        ...functions,
      ].join('\n');

      const extension = this.config.language === 'typescript' ? 'ts' : 'js';
      const typesFilePath = path.join(this.config.output, `api-types.${extension}`);
      const functionsFilePath = path.join(this.config.output, `api-functions.${extension}`);

      console.info(pc.blue('üìù Writing API types to:'), pc.cyan(typesFilePath));
      fs.writeFileSync(typesFilePath, typesOutput);

      console.info(pc.blue('üìù Writing API functions to:'), pc.cyan(functionsFilePath));
      fs.writeFileSync(functionsFilePath, functionsOutput);
    }

    const endpointCount = functions.length;
    console.info(
      pc.green(
        `‚ú® Generated ${endpointCount} API ${endpointCount === 1 ? 'endpoint' : 'endpoints'} successfully!`,
      ),
    );
  }
}
