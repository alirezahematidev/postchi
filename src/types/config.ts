import { z } from 'zod';

/**
 * Available programming languages for code generation
 */
export type Language = 'javascript' | 'typescript';

/**
 * Available HTTP request handlers
 */
export type RequestHandler = 'fetch' | 'axios';

/**
 * Output file strategy
 */
export type Strategy = 'single-file' | 'multi-file';

/**
 * Configuration interface for Postchi
 */
export interface PostchiConfig {
  /**
   * Path to the Postman collection JSON file
   */
  input?: string;

  /**
   * Output directory for generated files
   */
  output?: string;

  /**
   * Programming language to generate
   * @default 'typescript'
   */
  language?: Language;

  /**
   * HTTP request handler to use
   * @default 'fetch'
   */
  requestHandler?: RequestHandler;

  /**
   * File generation strategy
   * @default 'single-file'
   */
  strategy?: Strategy;
}

/**
 * Default configuration
 */
export const defaultConfig: Required<PostchiConfig> = {
  input: '',
  output: './src/api',
  language: 'typescript',
  requestHandler: 'fetch',
  strategy: 'single-file',
};

export const ConfigSchema = z
  .object({
    input: z.string().optional(),
    output: z.string().optional().default('./src/api'),
    language: z.enum(['javascript', 'typescript']).optional().default('typescript'),
    requestHandler: z.enum(['fetch', 'axios']).optional(),
    strategy: z.enum(['single-file', 'multi-file']).optional(),
  })
  .strict()
  .refine(
    (data: Partial<PostchiConfig>) => data.input !== undefined || process.env.NODE_ENV === 'test',
    {
      message: 'Input file path is required. You must specify a Postman collection JSON file.',
      path: ['input'],
    },
  );
