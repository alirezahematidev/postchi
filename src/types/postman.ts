export interface PostmanCollection {
  info: {
    name: string;
    schema: string;
  };
  item: PostmanItem[];
}

export interface PostmanItem {
  name: string;
  item?: PostmanItem[];
  request?: PostmanRequest;
  response?: PostmanResponse[];
}

export interface PostmanRequest {
  method: string;
  header: PostmanHeader[];
  url: PostmanUrl;
  body?: PostmanBody;
  description?: string;
}

export interface PostmanHeader {
  key: string;
  value: string;
  type: string;
}

export interface PostmanUrl {
  raw: string;
  protocol?: string;
  host: string[];
  path: string[];
  query?: PostmanQuery[];
}

export interface PostmanQuery {
  key: string;
  value: string;
}

export interface PostmanBody {
  mode: string;
  raw?: string;
  options?: {
    raw: {
      language: string;
    };
  };
}

export interface PostmanResponse {
  name: string;
  originalRequest: PostmanRequest;
  status: string;
  code: number;
  header: PostmanHeader[];
  body?: string;
}
