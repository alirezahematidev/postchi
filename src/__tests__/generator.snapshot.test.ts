import fs from 'fs';
import path from 'path';
import { ApiGenerator } from '../generator';
import { PostchiConfig } from '../types/config';

// Mock fs module
jest.mock('fs', () => ({
  ...jest.requireActual('fs'),
  readFileSync: jest.fn(),
  existsSync: jest.fn(),
  mkdirSync: jest.fn(),
  writeFileSync: jest.fn(),
}));

describe('ApiGenerator Snapshots', () => {
  const sampleCollectionPath = path.join(__dirname, 'fixtures', 'sample-collection.json');
  const complexCollectionPath = path.join(__dirname, 'fixtures', 'complex-collection.json');
  const outputDir = './test-output';

  // Default configuration
  const defaultConfig: Required<PostchiConfig> = {
    input: sampleCollectionPath,
    output: outputDir,
    language: 'typescript',
    requestHandler: 'fetch',
    strategy: 'single-file',
  };

  beforeEach(() => {
    // Reset all mocks before each test
    jest.clearAllMocks();

    // Mock fs.existsSync to return false (directory doesn't exist)
    (fs.existsSync as jest.Mock).mockReturnValue(false);
  });

  test('should generate expected API client for sample collection', () => {
    // Mock fs.readFileSync to return our sample collection
    (fs.readFileSync as jest.Mock).mockImplementation((filePath) => {
      if (filePath === sampleCollectionPath) {
        return JSON.stringify(require('./fixtures/sample-collection.json'));
      }
      return '';
    });

    const generator = new ApiGenerator(sampleCollectionPath, defaultConfig);
    generator.generate();

    // Check if writeFileSync was called
    expect(fs.writeFileSync).toHaveBeenCalled();

    // Get the generated content
    const mockCalls = (fs.writeFileSync as jest.Mock).mock.calls;
    expect(mockCalls.length).toBeGreaterThan(0);
    const generatedContent = mockCalls[0][1];

    // Match against snapshot
    expect(generatedContent).toMatchSnapshot('sample-collection-api-client');
  });

  test('should generate expected API client for complex collection', () => {
    // Create a complex collection
    const complexCollection = {
      info: {
        name: 'Complex API',
        schema: 'https://schema.getpostman.com/json/collection/v2.1.0/collection.json',
      },
      item: [
        {
          name: 'Authentication',
          item: [
            {
              name: 'Login',
              request: {
                method: 'POST',
                header: [
                  {
                    key: 'Content-Type',
                    value: 'application/json',
                    type: 'text',
                  },
                ],
                url: {
                  raw: 'https://api.example.com/auth/login',
                  protocol: 'https',
                  host: ['api', 'example', 'com'],
                  path: ['auth', 'login'],
                },
                body: {
                  mode: 'raw',
                  raw: JSON.stringify({
                    email: 'user@example.com',
                    password: 'password123',
                    rememberMe: true,
                  }),
                  options: {
                    raw: {
                      language: 'json',
                    },
                  },
                },
              },
            },
            {
              name: 'Refresh Token',
              request: {
                method: 'POST',
                header: [
                  {
                    key: 'Content-Type',
                    value: 'application/json',
                    type: 'text',
                  },
                  {
                    key: 'Authorization',
                    value: 'Bearer {{refreshToken}}',
                    type: 'text',
                  },
                ],
                url: {
                  raw: 'https://api.example.com/auth/refresh',
                  protocol: 'https',
                  host: ['api', 'example', 'com'],
                  path: ['auth', 'refresh'],
                },
              },
            },
          ],
        },
      ],
    };

    // Mock fs.readFileSync to return our complex collection
    (fs.readFileSync as jest.Mock).mockImplementation(() => JSON.stringify(complexCollection));

    const generator = new ApiGenerator(complexCollectionPath, defaultConfig);
    generator.generate();

    // Check if writeFileSync was called
    expect(fs.writeFileSync).toHaveBeenCalled();

    // Get the generated content
    const mockCalls = (fs.writeFileSync as jest.Mock).mock.calls;
    expect(mockCalls.length).toBeGreaterThan(0);
    const generatedContent = mockCalls[0][1];

    // Match against snapshot
    expect(generatedContent).toMatchSnapshot('complex-collection-api-client');
  });

  test('should generate expected API client for collection with query parameters', () => {
    // Create a collection with query parameters
    const queryParamsCollection = {
      info: {
        name: 'Query Params API',
        schema: 'https://schema.getpostman.com/json/collection/v2.1.0/collection.json',
      },
      item: [
        {
          name: 'List Users',
          request: {
            method: 'GET',
            header: [],
            url: {
              raw: 'https://api.example.com/users?page=1&limit=10&sort=name&order=asc',
              protocol: 'https',
              host: ['api', 'example', 'com'],
              path: ['users'],
              query: [
                {
                  key: 'page',
                  value: '1',
                },
                {
                  key: 'limit',
                  value: '10',
                },
                {
                  key: 'sort',
                  value: 'name',
                },
                {
                  key: 'order',
                  value: 'asc',
                },
              ],
            },
          },
        },
      ],
    };

    // Mock fs.readFileSync to return our query params collection
    (fs.readFileSync as jest.Mock).mockImplementation(() => JSON.stringify(queryParamsCollection));

    const generator = new ApiGenerator(sampleCollectionPath, defaultConfig);
    generator.generate();

    // Check if writeFileSync was called
    expect(fs.writeFileSync).toHaveBeenCalled();

    // Get the generated content
    const mockCalls = (fs.writeFileSync as jest.Mock).mock.calls;
    expect(mockCalls.length).toBeGreaterThan(0);
    const generatedContent = mockCalls[0][1];

    // Match against snapshot
    expect(generatedContent).toMatchSnapshot('query-params-api-client');
  });

  test('should generate expected API client for different HTTP methods', () => {
    // Create a collection with different HTTP methods
    const httpMethodsCollection = {
      info: {
        name: 'HTTP Methods API',
        schema: 'https://schema.getpostman.com/json/collection/v2.1.0/collection.json',
      },
      item: [
        {
          name: 'GET Request',
          request: {
            method: 'GET',
            header: [],
            url: {
              raw: 'https://api.example.com/resources/123',
              protocol: 'https',
              host: ['api', 'example', 'com'],
              path: ['resources', '123'],
            },
          },
        },
        {
          name: 'POST Request',
          request: {
            method: 'POST',
            header: [
              {
                key: 'Content-Type',
                value: 'application/json',
                type: 'text',
              },
            ],
            url: {
              raw: 'https://api.example.com/resources',
              protocol: 'https',
              host: ['api', 'example', 'com'],
              path: ['resources'],
            },
            body: {
              mode: 'raw',
              raw: JSON.stringify({
                name: 'New Resource',
                description: 'This is a new resource',
              }),
              options: {
                raw: {
                  language: 'json',
                },
              },
            },
          },
        },
      ],
    };

    // Mock fs.readFileSync to return our HTTP methods collection
    (fs.readFileSync as jest.Mock).mockImplementation(() => JSON.stringify(httpMethodsCollection));

    const generator = new ApiGenerator(sampleCollectionPath, defaultConfig);
    generator.generate();

    // Check if writeFileSync was called
    expect(fs.writeFileSync).toHaveBeenCalled();

    // Get the generated content
    const mockCalls = (fs.writeFileSync as jest.Mock).mock.calls;
    expect(mockCalls.length).toBeGreaterThan(0);
    const generatedContent = mockCalls[0][1];

    // Match against snapshot
    expect(generatedContent).toMatchSnapshot('http-methods-api-client');
  });

  test('should generate different output based on configuration', () => {
    // Mock fs.readFileSync to return our sample collection
    (fs.readFileSync as jest.Mock).mockImplementation((filePath) => {
      if (filePath === sampleCollectionPath) {
        return JSON.stringify(require('./fixtures/sample-collection.json'));
      }
      return '';
    });

    // Test JavaScript output
    const jsConfig: Required<PostchiConfig> = {
      ...defaultConfig,
      language: 'javascript',
    };

    const generator = new ApiGenerator(sampleCollectionPath, jsConfig);
    generator.generate();

    // Check if writeFileSync was called
    expect(fs.writeFileSync).toHaveBeenCalled();

    // Get the generated content
    const mockCalls = (fs.writeFileSync as jest.Mock).mock.calls;
    expect(mockCalls.length).toBeGreaterThan(0);
    const generatedContent = mockCalls[0][1];

    // Match against snapshot
    expect(generatedContent).toMatchSnapshot('javascript-api-client');

    // Reset mocks
    jest.clearAllMocks();

    // Test axios output
    const axiosConfig: Required<PostchiConfig> = {
      ...defaultConfig,
      requestHandler: 'axios',
    };

    const axiosGenerator = new ApiGenerator(sampleCollectionPath, axiosConfig);
    axiosGenerator.generate();

    // Check if writeFileSync was called
    expect(fs.writeFileSync).toHaveBeenCalled();

    // Get the generated content
    const mockCallsAxios = (fs.writeFileSync as jest.Mock).mock.calls;
    expect(mockCallsAxios.length).toBeGreaterThan(0);
    const generatedContentAxios = mockCallsAxios[0][1];

    // Match against snapshot
    expect(generatedContentAxios).toMatchSnapshot('axios-api-client');

    // Reset mocks
    jest.clearAllMocks();

    // Test multi-file strategy
    const multiFileConfig: Required<PostchiConfig> = {
      ...defaultConfig,
      strategy: 'multi-file',
    };

    const multiFileGenerator = new ApiGenerator(sampleCollectionPath, multiFileConfig);
    multiFileGenerator.generate();

    // Check if writeFileSync was called
    expect(fs.writeFileSync).toHaveBeenCalledTimes(2);

    // Get the generated content for types file
    const mockCallsTypes = (fs.writeFileSync as jest.Mock).mock.calls;
    expect(mockCallsTypes.length).toBeGreaterThan(0);
    const typesContent = mockCallsTypes[0][1];

    // Get the generated content for functions file
    const functionsContent = mockCallsTypes[1][1];

    // Match against snapshots
    expect(typesContent).toMatchSnapshot('multi-file-api-types');
    expect(functionsContent).toMatchSnapshot('multi-file-api-functions');
  });
});
