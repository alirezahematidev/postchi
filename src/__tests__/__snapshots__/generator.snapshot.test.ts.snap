// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiGenerator Snapshots should generate different output based on configuration: axios-api-client 1`] = `
"// Generated API client
// Do not modify this file manually

import axios from 'axios';

export interface HttpsApiExampleComUsersUserIdRequest {
}


export interface HttpsApiExampleComUsersRequest {
  body:   {
    "name": "John Doe",
    "email": "john@example.com",
    "age": 30
  };
}


export interface HttpsApiExampleComProductsCategoryElectronicsLimit10Request {
  category: string;
  limit: string;
}


export async function get_user(): Promise<Response> {
  const url = new URL('https://api.example.com/users/{{userId}}');
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return axios({
    method: 'get',
    url: url.toString(),
    headers: Object.fromEntries(headers.entries()),
  });
}


export async function create_user(body: HttpsApiExampleComUsersRequest['body']): Promise<Response> {
  const url = new URL('https://api.example.com/users');
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return axios({
    method: 'post',
    url: url.toString(),
    headers: Object.fromEntries(headers.entries()),
    data: body,
  });
}


export async function search_products(params: HttpsApiExampleComProductsCategoryElectronicsLimit10Request): Promise<Response> {
  const url = new URL('https://api.example.com/products?category=electronics&limit=10');
  Object.entries(params).forEach(([key, value]) => {
    url.searchParams.append(key, value);
  });
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return axios({
    method: 'get',
    url: url.toString(),
    headers: Object.fromEntries(headers.entries()),
  });
}

"
`;

exports[`ApiGenerator Snapshots should generate different output based on configuration: javascript-api-client 1`] = `
"// Generated API client
// Do not modify this file manually


/**
 * @typedef {Object} HttpsApiExampleComUsersUserIdRequest
 */


/**
 * @typedef {Object} HttpsApiExampleComUsersRequest
 * @property {Object} body
 * @property {string} body.name
 * @property {string} body.email
 * @property {number} body.age
 */


/**
 * @typedef {Object} HttpsApiExampleComProductsCategoryElectronicsLimit10Request
 * @property {string} category
 * @property {string} limit
 */


/**
 * @returns {Promise<Response>}
 */
export async function get_user() {
  const url = new URL('https://api.example.com/users/{{userId}}');
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return fetch(url.toString(), {
    method: 'GET',
    headers,
  });
}


/**
 * @param {Object} body
 * @returns {Promise<Response>}
 */
export async function create_user(body) {
  const url = new URL('https://api.example.com/users');
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return fetch(url.toString(), {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
  });
}


/**
 * @param {HttpsApiExampleComProductsCategoryElectronicsLimit10Request} params
 * @returns {Promise<Response>}
 */
export async function search_products(params) {
  const url = new URL('https://api.example.com/products?category=electronics&limit=10');
  Object.entries(params).forEach(([key, value]) => {
    url.searchParams.append(key, value);
  });
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return fetch(url.toString(), {
    method: 'GET',
    headers,
  });
}

"
`;

exports[`ApiGenerator Snapshots should generate different output based on configuration: multi-file-api-functions 1`] = `
"// Generated API functions
// Do not modify this file manually


import { HttpsApiExampleComUsersUserIdRequest, HttpsApiExampleComUsersRequest, HttpsApiExampleComProductsCategoryElectronicsLimit10Request } from './api-types';

export async function get_user(): Promise<Response> {
  const url = new URL('https://api.example.com/users/{{userId}}');
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return fetch(url.toString(), {
    method: 'GET',
    headers,
  });
}


export async function create_user(body: HttpsApiExampleComUsersRequest['body']): Promise<Response> {
  const url = new URL('https://api.example.com/users');
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return fetch(url.toString(), {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
  });
}


export async function search_products(params: HttpsApiExampleComProductsCategoryElectronicsLimit10Request): Promise<Response> {
  const url = new URL('https://api.example.com/products?category=electronics&limit=10');
  Object.entries(params).forEach(([key, value]) => {
    url.searchParams.append(key, value);
  });
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return fetch(url.toString(), {
    method: 'GET',
    headers,
  });
}

"
`;

exports[`ApiGenerator Snapshots should generate different output based on configuration: multi-file-api-types 1`] = `
"// Generated API types
// Do not modify this file manually

export interface HttpsApiExampleComUsersUserIdRequest {
}


export interface HttpsApiExampleComUsersRequest {
  body:   {
    "name": "John Doe",
    "email": "john@example.com",
    "age": 30
  };
}


export interface HttpsApiExampleComProductsCategoryElectronicsLimit10Request {
  category: string;
  limit: string;
}

"
`;

exports[`ApiGenerator Snapshots should generate expected API client for collection with query parameters: query-params-api-client 1`] = `
"// Generated API client
// Do not modify this file manually


export interface HttpsApiExampleComUsersPage1Limit10SortNameOrderAscRequest {
  page: string;
  limit: string;
  sort: string;
  order: string;
}


export async function list_users(params: HttpsApiExampleComUsersPage1Limit10SortNameOrderAscRequest): Promise<Response> {
  const url = new URL('https://api.example.com/users?page=1&limit=10&sort=name&order=asc');
  Object.entries(params).forEach(([key, value]) => {
    url.searchParams.append(key, value);
  });
  const headers = new Headers({
  });
  return fetch(url.toString(), {
    method: 'GET',
    headers,
  });
}

"
`;

exports[`ApiGenerator Snapshots should generate expected API client for complex collection: complex-collection-api-client 1`] = `
"// Generated API client
// Do not modify this file manually


export interface HttpsApiExampleComAuthLoginRequest {
  body:   {
    "email": "user@example.com",
    "password": "password123",
    "rememberMe": true
  };
}


export interface HttpsApiExampleComAuthRefreshRequest {
}


export async function login(body: HttpsApiExampleComAuthLoginRequest['body']): Promise<Response> {
  const url = new URL('https://api.example.com/auth/login');
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return fetch(url.toString(), {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
  });
}


export async function refresh_token(): Promise<Response> {
  const url = new URL('https://api.example.com/auth/refresh');
  const headers = new Headers({
    'Content-Type': 'application/json',
    'Authorization': 'Bearer {{refreshToken}}',
  });
  return fetch(url.toString(), {
    method: 'POST',
    headers,
  });
}

"
`;

exports[`ApiGenerator Snapshots should generate expected API client for different HTTP methods: http-methods-api-client 1`] = `
"// Generated API client
// Do not modify this file manually


export interface HttpsApiExampleComResources123Request {
}


export interface HttpsApiExampleComResourcesRequest {
  body:   {
    "name": "New Resource",
    "description": "This is a new resource"
  };
}


export async function get_request(): Promise<Response> {
  const url = new URL('https://api.example.com/resources/123');
  const headers = new Headers({
  });
  return fetch(url.toString(), {
    method: 'GET',
    headers,
  });
}


export async function post_request(body: HttpsApiExampleComResourcesRequest['body']): Promise<Response> {
  const url = new URL('https://api.example.com/resources');
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return fetch(url.toString(), {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
  });
}

"
`;

exports[`ApiGenerator Snapshots should generate expected API client for sample collection: sample-collection-api-client 1`] = `
"// Generated API client
// Do not modify this file manually


export interface HttpsApiExampleComUsersUserIdRequest {
}


export interface HttpsApiExampleComUsersRequest {
  body:   {
    "name": "John Doe",
    "email": "john@example.com",
    "age": 30
  };
}


export interface HttpsApiExampleComProductsCategoryElectronicsLimit10Request {
  category: string;
  limit: string;
}


export async function get_user(): Promise<Response> {
  const url = new URL('https://api.example.com/users/{{userId}}');
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return fetch(url.toString(), {
    method: 'GET',
    headers,
  });
}


export async function create_user(body: HttpsApiExampleComUsersRequest['body']): Promise<Response> {
  const url = new URL('https://api.example.com/users');
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return fetch(url.toString(), {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
  });
}


export async function search_products(params: HttpsApiExampleComProductsCategoryElectronicsLimit10Request): Promise<Response> {
  const url = new URL('https://api.example.com/products?category=electronics&limit=10');
  Object.entries(params).forEach(([key, value]) => {
    url.searchParams.append(key, value);
  });
  const headers = new Headers({
    'Content-Type': 'application/json',
  });
  return fetch(url.toString(), {
    method: 'GET',
    headers,
  });
}

"
`;
